src/java/main/org/apache/zookeeper/ZooKeeper.java

```diff
@@ -1325,6 +1325,23 @@ public synchronized void close() throws InterruptedException {
     }
 
     /**
+     * Close this client object as the {@link #close() } method.
+     * This method will wait for internal resources to be released.
+     *
+     * @param waitForShutdownTimeoutMs timeout (in milliseconds) to wait for resources to be released.
+     * Use zero or a negative value to skip the wait
+     * @throws InterruptedException
+     * @return true if waitForShutdownTimeout is greater than zero and all of the resources have been released
+     */
+    public boolean close(int waitForShutdownTimeoutMs) throws InterruptedException {
```

Comment:
synchronization is in the real close() method. IMHO If I will add the synch here I think we will introduce unpredictable side effects